File: jar-shell/src/main/java/com/beancounter/shell/ingest/IngestionFactory.java
package com.beancounter.shell.ingest;

import com.beancounter.shell.csv.CsvIngester;
import com.beancounter.shell.google.GoogleConfig;
import com.beancounter.shell.google.SheetIngester;
import java.util.HashMap;
import java.util.Map;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.stereotype.Service;

@Service
public class IngestionFactory {
  private Map<String, Ingester> ingesterMap = new HashMap<>();

  @Autowired
  void setCsvIngester(CsvIngester csvIngester) {
    add("CSV", csvIngester);
  }

  @Autowired(required = false)
  @ConditionalOnBean(GoogleConfig.class)
  void setSheetIngester(SheetIngester sheetIngester) {
    add("GSHEET", sheetIngester);
  }


  public Ingester getIngester(IngestionRequest ingestionRequest) {
    return ingesterMap.get(ingestionRequest.getType());
  }

  public void add(String key, Ingester ingester) {
    ingesterMap.put(key.toUpperCase(), ingester);
  }
}

File: jar-shell/src/main/kotlin/com/beancounter/shell/ingest/IngestionFactory.kt
package com.beancounter.shell.ingest

import com.beancounter.common.exception.SystemException
import com.beancounter.shell.csv.CsvIngester
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.Locale

/**
 * All known Ingestion request handlers.
 */
@Service
class IngestionFactory {
    private val ingesterMap: MutableMap<String, Ingester> = HashMap()

    @Autowired
    fun setCsvIngester(csvIngester: CsvIngester) {
        add("CSV", csvIngester)
    }

    fun getIngester(ingestionRequest: IngestionRequest): Ingester {
        return ingesterMap[ingestionRequest.reader.uppercase(Locale.getDefault())]
            ?: throw SystemException("Unable to resolve ingestor for ${ingestionRequest.reader}")
    }

    fun add(key: String, ingester: Ingester) {
        ingesterMap[key.uppercase(Locale.getDefault())] = ingester
    }
}
